### PR TITLE
:ambulance: Restore legacy file deletion, handle failed books, add comments

### PR DESCRIPTION
**Pull request by tyler-dot-earth**
tl;dr deletion logic changes in #67 had oversight in handling existing deletions which caused unintended re-syncs.

This PR:
- Generally restores the previous method of determining deleted files
- Adds additional comments
- Specifically handles failed books separately of books to refresh (deleted books)

Relevant thread: https://discord.com/channels/886992134505398314/894996177790046228/1283552094296608871

## mental model

HOW IT WORKED BEFORE #67 when a user deleted a file
1. file deleted
2. file path/id removed from `booksIDsMap`
3. book ID added to `booksToRefresh`
4. when syncing, `booksToRefresh` only used when `resync deleted files` enabled
ADDITIONALLY:
- `booksToRefresh` was shared for both deleted and failed book ID collection
- ... which means that failed books would *never* resync if user did not have `resync deleted books` disabled (pre-existing bug)

HOW IT WORKED AFTER #67
1. file deleted
2. file path/id **remains** in `booksIDsMap`
3. book id **not** added to `booksToRefresh`
4. when syncing, (i) `booksToRefresh` **always** used when syncing, (ii) resyncing deleted files determined solely by delta between filesystem & `booksIDsMap`

BUG CAUSLITY:
- (4)(i) was bad logic rework; did not account for legacy book deletion behavior
- `booksToRefresh` list of IDs should only have been used when `resync deleted books` enabled
- additionaly, more work needed to be put into place to handle failed vs deleted books (as the only remaining use of `booksToRefresh` in v2.1 was for failed books)
- another case: if user had bad `path: bookid` in `booksIDsMap` (mismatch between local and db filenames), `refresh_book_export` would be hit every single sync

## repro/testing/validation steps

- sync readwise once
- disable "resync deleted files"
- disable the readwise plugin
- open readwise plugin `data.json`
- pick any book from booksIDsMap
  - copy its ID into `booksToRefresh` array (this replicates pre-#67 deletion behavior)
  - delete the line from `booksIDsMap` (this replicates pre-#67 deletion behavior)
  - save the `data.json` file
  - delete that file from obsidian / your filesystem
- for good measure, reload obsidian workspace (ctrl p > reload app without saving)
- re-enable readwise plugin
- initiate a readwise sync (in plugins, or command palette) if it did not do it upon the plugin being enabled
- after readwise sync completes, verify `data.json`:
  - [ ] the ID you placed in `booksToRefresh` should still be there
  - [ ] the line you deleted from `booksIDsMap` should still NOT be there
  - [ ] the file you deleted from obsidian / your filesystem should still NOT be there
- create a duplicate entry in your `data.json`:
  - edit your `data.json`
  - duplicate one line from `booksIDsMap` but change the **key** (filepath) to something else like `"xxx CHANGED.md": 1234`
  - ensure `resync deleted files` is on
  - re-sync
  - [ ] the non-existant path/file in `booksIDsMap` that you created shouldn't cause `refresh_book_export` to be hit in network calls (because code no longer looks at filesystem delta)

**additional, more sync testing steps below for thoroughness**

> [!WARNING]  
> perform these in listed order

- [ ] confirm clean plugin install syncs notes as expected (this isn't automatic; you must click "initiate sync" or reload the workspace after authenticating)

- **enable `Resync deleted files`**
- delete a synced file
- use command palette to `Readwise Official: Sync your data now`
- [ ] confirm deleted file **is** re-synced

- delete a synced file
- open plugin settings and click `Initiate sync`
- [ ] confirm deleted file **is** re-synced

- open a synced file
- use command palette to `Readwise Official: Delete and reimport this document`
- [ ] confirm file is deleted and **is** re-synced

- delete synced file
- use command palette to `Reload app without saving` (or re-open Obsidian)
- [ ] confirm deleted file **is** re-synced

- **disable `Resync deleted files`**
- delete a synced file
- use command palette to `Readwise Official: Sync your data now`
- [ ] confirm file is **NOT** re-synced

- delete a synced file
- open plugin settings and click `Initiate sync`
- [ ] confirm file is **NOT** re-synced

- delete synced file
- use command palette to `Reload app without saving` (or re-open Obsidian)
- [ ] confirm deleted file **isn't** re-synced

- open a synced file
- use command palette to `Readwise Official: Delete and reimport this document`
- [ ] confirm file **is** re-synced

- delete synced file
- **enable `Resync deleted files`**
- [ ] confirm deleted file **is** re-synced

- and finally, for good measure:
- [ ] ensure adding new highlights continues to work as expected
- [ ] ensure syncing on interval continues to work as expected

### CODE REVIEW

**Comment by tyler-dot-earth** [comment]
(draft because i'm still verifying logic/behavior)

File: src/main.ts [hunk 1]
```diff
@@ -55,9 +55,13 @@ interface ReadwisePluginSettings {
   /** Should get any deleted books */
   refreshBooks: boolean,
 
-  /** Queue of books to refresh */
+  /** Queue of books to refresh.
+   * @deprecated LEGACY: only exists to preserve books that were deleted prior to v2.1.0 */
   booksToRefresh: Array<string>;
 
+  /** retryFailedBooks: queue of books to retry because of previous failure */
+  failedBooks: Array<string>;
```
Comment (tyler-dot-earth) [inline src/main.ts:63]: handles:
1. new deletion approach from #67
2. pre-existing bug wherein failed books would never resync if user did not have resync deleted books disabled

File: src/main.ts [hunk 2]
```diff
@@ -361,16 +366,18 @@ export default class ReadwisePlugin extends Plugin {
...
           this.notice(`Readwise: error while writing ${processedFileName}: ${e}`, true, 4, true);
           if (bookID) {
             // handles case where user doesn't have `settings.refreshBooks` enabled
-            await this.addBookToRefresh(bookID);
+            await this.addToFailedBooks(bookID);
           }
           // communicate with readwise?
         }
       }
+      await this.saveSettings();
```
Comment (tyler-dot-earth) [inline src/main.ts:380]: batch save instead of saving each `this.removeX` call

File: src/main.ts [hunk 3]
```diff
@@ -430,7 +437,12 @@ export default class ReadwisePlugin extends Plugin {
   ) {
     if (!this.settings.token) return;
 
-    const targetBookIds = bookIds || this.settings.booksToRefresh;
+    let targetBookIds = bookIds || this.settings.failedBooks;
+
+    // handle legacy `booksToRefresh` setting; see tsdoc on bookssToRefresh
+    if (this.settings.refreshBooks) {
+      targetBookIds = targetBookIds || this.settings.booksToRefresh;
+    }
```
Comment (tyler-dot-earth) [inline src/main.ts:445]: only use `booksToRefresh` if `resync deleted files` is enabled (and no `targetBookIds` already provided)

Comment (tyler-dot-earth) [reply src/main.ts:445]: potential concern: if failed books continue to fail, they'll always end up in targetBookIds which means that booksToRefresh would *never* get used. :thinking:

File: src/main.ts [hunk 4]
```diff
@@ -485,20 +497,33 @@ export default class ReadwisePlugin extends Plugin {
     }
   }
 
-  async addBookToRefresh(bookId: string) {
-    let booksToRefresh = this.settings.booksToRefresh;
-    booksToRefresh.push(bookId);
-    console.log(`Readwise Official plugin: added book id ${bookId} to refresh later`);
-    this.settings.booksToRefresh = booksToRefresh;
+  async addToFailedBooks(bookId: string) {
+    let failedBooks = [...(this.settings.failedBooks || DEFAULT_SETTINGS.failedBooks)];
```
Comment (tyler-dot-earth) [inline src/main.ts:501]: don't mutate the original `this.settings.failedBooks`

File: src/main.ts (hunk 3 above)
Comment (TristanH) [reply src/main.ts:445]: yeah.. could you just append `this.settings.booksToRefresh` to the end of targetBookIds?

File: src/main.ts [hunk 5]
```diff
@@ -361,16 +366,18 @@ export default class ReadwisePlugin extends Plugin {
...
           await this.fs.write(originalName, contentToSave);
           await this.removeBooksFromRefresh([bookID]);
+          await this.removeBookFromFailedBooks([bookID]);
         } catch (e) {
           console.log(`Readwise Official plugin: error writing ${processedFileName}:`, e);
           this.notice(`Readwise: error while writing ${processedFileName}: ${e}`, true, 4, true);
           if (bookID) {
             // handles case where user doesn't have `settings.refreshBooks` enabled
-            await this.addBookToRefresh(bookID);
+            await this.addToFailedBooks(bookID);
```
Comment (TristanH) [inline src/main.ts:375]: to be safe, shouldn't this catch really only be for the `fs.write` call? it makes my head spin thinking a little bit about `this.removeBookFromFailedBooks` somehow failing, causing us to add that same book to the failedbooks again

File: src/main.ts (hunk 3 above)
Comment (tyler-dot-earth) [reply src/main.ts:446]: i think the odd-man-out case would be that the `delete and reimport` command may end up importing *more* than the single book you expect. gimme a few mins to think.

Comment (tyler-dot-earth) [reply src/main.ts]: i believe i've solved this concern.

File: src/main.ts [hunk 6]
```diff
@@ -360,17 +367,20 @@ export default class ReadwisePlugin extends Plugin {
...
           }
           await this.fs.write(originalName, contentToSave);
-          await this.removeBooksFromRefresh([bookID]);
         } catch (e) {
           console.log(`Readwise Official plugin: error writing ${processedFileName}:`, e);
           this.notice(`Readwise: error while writing ${processedFileName}: ${e}`, true, 4, true);
           if (bookID) {
             // handles case where user doesn't have `settings.refreshBooks` enabled
-            await this.addBookToRefresh(bookID);
+            await this.addToFailedBooks(bookID);
```
Comment (TristanH) [inline src/main.ts:375]: ummm i think you need to exit early here right? otherwise the bookId will get removed from the list rigth after being added on line 381?

Comment (TristanH) [reply src/main.ts:375]: (this wasn't the case until i suggested moving those remove calls outside the try/catch 😢

Comment (tyler-dot-earth) [reply src/main.ts:375]: f1d67e2

Comment (tyler-dot-earth) [reply src/main.ts:375]: also aligned this method with the other add/remove to refreshbooks fns WRT how it saves: 3092e8f

File: src/main.ts [hunk 7]
```diff
@@ -430,38 +442,38 @@ export default class ReadwisePlugin extends Plugin {
...
+        ...this.settings.booksToRefresh,
+      ];
     }
 
     if (!targetBookIds.length) {
-      console.log('Readwise Official plugin: no targetBookIds, triggering check for other updates...');
-      await this.queueExport(null, null, auto);
+      console.log('Readwise Official plugin: no targetBookIds, checking for new highlights');
+      // no need to hit refresh_book_export;
+      // just check if there's new highlights from the server
```
Comment (TristanH) [inline src/main.ts:464]: nit: i'd leave a comment pointing out this is the "default" path if a user doesn't have any deleted files they need to resync
